File: backend/internal/api/admin/contest.go
package admin

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"image"
	_ "image/gif"
	_ "image/jpeg"
	_ "image/png"

	"connectrpc.com/connect"
	adminv1 "github.com/LuukBlankenstijn/loom/gen/go/admin/v1"
	_ "golang.org/x/image/bmp"
	_ "golang.org/x/image/tiff"
	_ "golang.org/x/image/webp"
	"google.golang.org/protobuf/types/known/emptypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// Gets the contest that is currently active, or the contest that will start next
func (a *adminHandler) GetNextContest(
	ctx context.Context,
	empty *emptypb.Empty,
) (*adminv1.Contest, error) {
	nextContest, err := a.contestRepo.GetNextContest(ctx)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, errors.New("failed to get next contest"))
	}
	if nextContest == nil {
		return nil, connect.NewError(connect.CodeNotFound, errors.New("next contest not found"))
	}

	var mapId *int32
	m, err := a.mapRepo.GetByContest(ctx, nextContest.Id)
	if err == nil && m != nil {
		val := int32(m.Id)
		mapId = &val
	}
	return &adminv1.Contest{
		Id:        nextContest.Id,
		Name:      nextContest.Name,
		StartTime: timestamppb.New(nextContest.StartTime),
		EndTime:   timestamppb.New(nextContest.EndTime),
		MapId:     mapId,
	}, nil
}

// Sets the wallpaper for some contest
func (a *adminHandler) SetWallpaper(
	ctx context.Context,
	request *adminv1.UploadWallpaperRequest,
) (*emptypb.Empty, error) {
	if len(request.ImageData) > 0 {
		mimeType, err := validateImageFormat(request.ImageData)
		if err != nil {
			return nil, connect.NewError(connect.CodeInvalidArgument, err)
		}
		if err := a.wallpaperRepo.SetWallpaperData(ctx, request.ContestId, request.ImageData, mimeType); err != nil {
			return nil, connect.NewError(connect.CodeInternal, err)
		}
		return &emptypb.Empty{}, nil
	}
	if err := a.wallpaperRepo.DeleteWallpaper(ctx, request.ContestId); err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return &emptypb.Empty{}, nil
}

func (a *adminHandler) SetWallpaperTextColor(
	ctx context.Context,
	request *adminv1.SetWallpaperTextColorRequest,
) (*emptypb.Empty, error) {
	if err := a.wallpaperRepo.SetWallpaperTextColor(ctx, request.ContestId, request.Color); err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return &emptypb.Empty{}, nil
}

func (a *adminHandler) GetWallpaper(
	ctx context.Context,
	request *adminv1.GetWallpaperRequest,
) (*adminv1.WallpaperResponse, error) {
	// if id is nil, try to get the first active contest
	if request.ContestId == nil {
		contest, err := a.contestRepo.GetNextContest(ctx)
		if err != nil {
			return nil, connect.NewError(
				connect.CodeInternal,
				fmt.Errorf("failed to get next context to get wallpaper for"),
			)
		}
		if contest == nil {
			return &adminv1.WallpaperResponse{ImageData: []byte{}}, nil
		}
		request.ContestId = &contest.Id
	}
	wallpaper, err := a.wallpaperRepo.GetWallpaper(ctx, *request.ContestId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	if wallpaper == nil {
		return &adminv1.WallpaperResponse{ImageData: []byte{}}, nil
	}
	return &adminv1.WallpaperResponse{ImageData: wallpaper.Data, Color: &wallpaper.TextColor}, nil
}

// Sets the map for some contest, does not error when either does not exist
func (a *adminHandler) SetMap(
	ctx context.Context,
	request *adminv1.SetMapRequest,
) (*emptypb.Empty, error) {
	err := a.mapRepo.SetMap(ctx, int(request.MapId), request.ContestId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return &emptypb.Empty{}, nil
}

var formatToMIME = map[string]string{
	"jpeg": "image/jpeg",
	"png":  "image/png",
	"gif":  "image/gif",
	"webp": "image/webp",
	"tiff": "image/tiff",
	"bmp":  "image/bmp",
}

func validateImageFormat(data []byte) (string, error) {
	if len(data) == 0 {
		return "", fmt.Errorf("empty image data")
	}
	const maxSize = 10 * 1024 * 1024
	if len(data) > maxSize {
		return "", fmt.Errorf("image too large: %d bytes (max %d bytes)", len(data), maxSize)
	}
	config, format, err := image.DecodeConfig(bytes.NewReader(data))
	if err != nil {
		return "", fmt.Errorf("invalid or corrupted image: %w", err)
	}
	mimeType, ok := formatToMIME[format]
	if !ok {
		return "", fmt.Errorf("unsupported image format: %s", format)
	}
	if config.Width < 1 || config.Height < 1 {
		return "", fmt.Errorf("invalid image dimensions: %dx%d", config.Width, config.Height)
	}
	return mimeType, nil
}

File: backend/src/api/admin/contest.rs
use derive_more::derive::Constructor;
use loom_proto_bridge::IntoProto;
use loom_rpc::admin::v1::{self as pb, contest_service_server::ContestService};
use std::sync::Arc;
use tonic::{Request, Response, Status};

use crate::domain::{ContestRepository, MapRepository, Orchestrator};

#[derive(Constructor)]
pub struct ContestHandler {
    contest_repo: Arc<dyn ContestRepository>,
    map_repo: Arc<dyn MapRepository>,
    orchestrator: Arc<dyn Orchestrator>,
}

#[tonic::async_trait]
impl ContestService for ContestHandler {
    async fn get_next_contest(
        &self,
        _request: Request<()>,
    ) -> Result<Response<pb::Contest>, Status> {
        let contest = self
            .contest_repo
            .get_next_contest()
            .await?
            .ok_or_else(|| Status::not_found("no upcoming contest"))?;

        let map_id = self
            .map_repo
            .get_by_contest(&contest.id)
            .await
            .ok()
            .flatten()
            .map(|m| m.id);

        let mut pb_contest: pb::Contest = contest.into_proto();
        pb_contest.map_id = map_id;
        Ok(Response::new(pb_contest))
    }

    async fn set_wallpaper(
        &self,
        request: Request<pb::UploadWallpaperRequest>,
    ) -> Result<Response<()>, Status> {
        let req = request.into_inner();
        match req.image_data.filter(|d| !d.is_empty()) {
            Some(data) => {
                let mime_type = validate_image(&data).map_err(Status::invalid_argument)?;
                self.contest_repo
                    .set_wallpaper(&req.contest_id, &data, mime_type)
                    .await?;
            }
            None => {
                self.contest_repo.delete_wallpaper(&req.contest_id).await?;
            }
        }
        self.orchestrator.sync_stations(&[]);
        Ok(Response::new(()))
    }

    async fn set_wallpaper_text_color(
        &self,
        request: Request<pb::SetWallpaperTextColorRequest>,
    ) -> Result<Response<()>, Status> {
        let req = request.into_inner();
        self.contest_repo
            .set_wallpaper_text_color(&req.contest_id, &req.color)
            .await?;
        // update stations
        self.orchestrator.sync_stations(&[]);
        Ok(Response::new(()))
    }

    async fn set_map(&self, request: Request<pb::SetMapRequest>) -> Result<Response<()>, Status> {
        let req = request.into_inner();
        self.contest_repo
            .set_map(&req.contest_id, req.map_id)
            .await?;
        Ok(Response::new(()))
    }
}

fn validate_image(data: &[u8]) -> Result<&'static str, String> {
    use image::ImageFormat;
    use std::io::Cursor;

    let format = image::guess_format(data).map_err(|_| "unsupported image format".to_string())?;

    let reader = image::ImageReader::new(Cursor::new(data))
        .with_guessed_format()
        .map_err(|e| format!("failed to read image: {e}"))?;

    reader
        .into_dimensions()
        .map_err(|e| format!("invalid image data: {e}"))?;

    let mime = match format {
        ImageFormat::Png => "image/png",
        ImageFormat::Jpeg => "image/jpeg",
        ImageFormat::Gif => "image/gif",
        ImageFormat::WebP => "image/webp",
        ImageFormat::Bmp => "image/bmp",
        ImageFormat::Tiff => "image/tiff",
        _ => return Err("unsupported image format".to_string()),
    };

    Ok(mime)
}
